packages/client/src/async-completion-client.ts

```diff
@@ -0,0 +1,250 @@
+   */
+  async fail(taskToken: Uint8Array, err: unknown): Promise<void>;
+  /**
+   * Fail an Activity by full ID
+   */
+  async fail(fullActivityId: FullActivityId, err: unknown): Promise<void>;
+
+  async fail(taskTokenOrFullActivityId: Uint8Array | FullActivityId, err: unknown): Promise<void> {
+    try {
+      if (taskTokenOrFullActivityId instanceof Uint8Array) {
+        await this.service.respondActivityTaskFailed({
+          identity: this.options.identity,
+          namespace: this.options.namespace,
+          taskToken: taskTokenOrFullActivityId,
+          failure: await errorToFailure(ensureTemporalFailure(err), this.options.dataConverter),
+        });
+      } else {
+        await this.service.respondActivityTaskFailedById({
+          identity: this.options.identity,
+          namespace: this.options.namespace,
+          ...taskTokenOrFullActivityId,
+          failure: await errorToFailure(err, this.options.dataConverter),
+        });
+      }
+    } catch (err) {
+      this.handleError(err);
+    }
+  }
+
+  /**
+   * Report Activity cancellation by task token
+   */
+  reportCancellation(taskToken: Uint8Array, details?: unknown): Promise<void>;
+  /**
+   * Report Activity cancellation by full ID
+   */
+  reportCancellation(fullActivityId: FullActivityId, details?: unknown): Promise<void>;
+
+  async reportCancellation(taskTokenOrFullActivityId: Uint8Array | FullActivityId, details?: unknown): Promise<void> {
+    try {
+      if (taskTokenOrFullActivityId instanceof Uint8Array) {
+        await this.service.respondActivityTaskCanceled({
+          identity: this.options.identity,
+          namespace: this.options.namespace,
+          taskToken: taskTokenOrFullActivityId,
+          details: { payloads: await this.options.dataConverter.toPayloads(details) },
+        });
+      } else {
+        await this.service.respondActivityTaskCanceledById({
+          identity: this.options.identity,
+          namespace: this.options.namespace,
+          ...taskTokenOrFullActivityId,
+          details: { payloads: await this.options.dataConverter.toPayloads(details) },
+        });
+      }
+    } catch (err) {
+      this.handleError(err);
+    }
+  }
+
+  /**
+   * Send Activity heartbeat by task token
+   */
+  heartbeat(taskToken: Uint8Array, details?: unknown): Promise<void>;
+  /**
+   * Send Activity heartbeat by full ID
+   */
+  heartbeat(fullActivityId: FullActivityId, details?: unknown): Promise<void>;
+
+  async heartbeat(taskTokenOrFullActivityId: Uint8Array | FullActivityId, details?: unknown): Promise<void> {
+    try {
+      if (taskTokenOrFullActivityId instanceof Uint8Array) {
+        const { cancelRequested } = await this.service.recordActivityTaskHeartbeat({
+          identity: this.options.identity,
+          namespace: this.options.namespace,
+          taskToken: taskTokenOrFullActivityId,
+          details: { payloads: await this.options.dataConverter.toPayloads(details) },
```

Comment:
This struct could be lifted up and then just modified in the one branch

Reply:
sure, but I feel like this is ok, it still fits in a single line.

Reply:
I meant the whole thing, not just details